File: challenges/2020/09-encodingError/go/challenge/partOne.go
package challenge

import (
	"sort"
)

func PartOne(instr string) int {
	inputSlice := parse(instr)

	preambleLen := 25
	if len(inputSlice) < 30 { // This is for test inputs
		preambleLen = 5
	}

	pointer := preambleLen
	for pointer < len(inputSlice) {
		if !hasCombinations(inputSlice[pointer-preambleLen:pointer], inputSlice[pointer]) {
			return inputSlice[pointer]
		}
		pointer += 1
	}

	return 0
}

func hasCombinations(options []int, target int) bool {

	// Remove duplicate options
	{
		keys := make(map[int]bool)
		var new []int

		for _, entry := range options {
			if _, value := keys[entry]; !value {
				keys[entry] = true
				new = append(new, entry)
			}
		}

		options = new
	}

	sort.Ints(options) // sorts in place

	var lPtr int
	rPtr := len(options) - 1

	for lPtr < rPtr {
		v := options[lPtr] + options[rPtr]
		if v == target {
			return true
		} else if v < target {
			lPtr += 1
		} else {
			rPtr -= 1
		}
	}

	return false
}

File: challenges/2020/09-encodingError/python/partOne.py
from typing import Set

from common import *


def has_combinations(options: Set[int], target: int) -> bool:
    # Returns true of the options set has two values that sum to be the target, else returns false

    options = sorted(options)

    l_ptr = 0
    r_ptr = len(options) - 1

    while l_ptr < r_ptr:
        v = options[l_ptr] + options[r_ptr]
        if v == target:
            return True
        elif v < target:
            l_ptr += 1
        else:
            r_ptr -= 1

    return False


def partOne(instr: str) -> int:
    input_list = parse(instr)

    preamble_len = 25
    if len(input_list) < 30:  # This is for tests
        preamble_len = 5

    pointer = preamble_len
    while pointer < len(input_list):
        if not has_combinations(
            set(input_list[pointer - preamble_len : pointer]), input_list[pointer]
        ):
            return input_list[pointer]
        pointer += 1

    return 0
